src/thor/expansion_action.cc

```diff
@@ -0,0 +1,144 @@
+#include "thor/worker.h"
+
+#include "baldr/json.h"
+#include "baldr/rapidjson_utils.h"
+#include "midgard/constants.h"
+#include "midgard/logging.h"
+#include "midgard/polyline2.h"
+#include "midgard/util.h"
+
+using namespace rapidjson;
+using namespace valhalla::midgard;
+
+namespace {
+void FillDom(Document& dom, valhalla::Options::Action action) {
+  SetValueByPointer(dom, "/type", "FeatureCollection");
+  SetValueByPointer(dom, "/features/0/type", "Feature");
+  SetValueByPointer(dom, "/features/0/geometry/type", "MultiLineString");
+  SetValueByPointer(dom, "/features/0/geometry/coordinates", Value(kArrayType));
+  if (action == valhalla::Options::route) {
+    SetValueByPointer(dom, "/properties/algorithm", "none");
+    SetValueByPointer(dom, "/features/0/properties/edge_ids", Value(kArrayType));
+    SetValueByPointer(dom, "/features/0/properties/statuses", Value(kArrayType));
+  } else {
+    SetValueByPointer(dom, "/features/0/properties/durations", Value(kArrayType));
+    SetValueByPointer(dom, "/features/0/properties/distances", Value(kArrayType));
+    SetValueByPointer(dom, "/features/0/properties/costs", Value(kArrayType));
+  }
+}
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+std::string thor_worker_t::expansion(Api& request) {
+  // time this whole method and save that statistic
+  measure_scope_time(request, "thor_worker_t::expansion");
+
+  auto options = request.options();
+  auto exp_action = options.expansion_action();
+  // default generalization to ~ zoom level 15
+  float gen_factor = options.has_generalize() ? options.generalize() : 21.4f;
+
+  // default the expansion geojson so its easy to add to as we go
+  Document dom;
+  dom.SetObject();
+  FillDom(dom, exp_action);
+
+  // a lambda that the path algorithm can call to add stuff to the dom
+  // route and isochrone produce different GeoJSON properties
```

Comment:
does it have to lead to different properties though? we track the same values in edge labels regardless of the algorithm type. i didnt go back and look at bidirea* and all the places it fires the expansion callback but if we have the label then we could also pass the details that isochrones passes. i have to say im not a huge fan of the outputs differing between different algorithms. im totally not opposed to changing the format now though, as you mentioned elsewhere this api is not even in the docs yet so its ok to make changes to it (we can consider it beta haha).

maybe you could fill in the section in the docs that you made with the proposed output so it would be easier to see where different actions would differ? i had hoped that at the worst one algorithm might have less or more information than the other but the properties in the geojson wouldnt be wholly different. maybe that is the case, i didnt try it out yet.

Reply:
> i have to say im not a huge fan of the outputs differing between different algorithms.

me neither really.. but those arrays are not cheap in terms of output size (calculated it yesterday during jogging, think was roughly smth like 20% of the `geometry` size), so I'd prefer even more to not have `edge_ids` (only good for debugging isnt it?) & `statuses` (pointless, we're only tracking settled edges so far) in the isochrone output.

I let you decide if you want those additional attributes (dist, dur, cost) for the routing output. 

That being said, I think the `isochrone` action has more meaning than `route`, so it's more user-facing. Consequently I think it's fine if that's returning smth entirely different. 

We could also at least have the same keys in both responses, but then some empty arrays? On the other hand it's easier to add attributes down the road rather than removing any.

What do you think?

Reply:
right now it's:
- isochrones: `durations`, `distances`, `costs`
- route: `statuses`, `edge_ids`

I can include the isochrone ones to `route` as well?

Reply:
yeah i feel like why not all of them in both? maybe the api lets you specify which bits of information you care about getting back?

Reply:
I'll get back to this in a bit

Reply:
implemented this